### DISCUSSION TITLE
IPV4 in IPV6 channel - xfrm issue

### DISCUSSION THREAD

**Discussion by SagarGowda89** [strongswan/strongswan#836] [Q&A]
Hi

My setup is 
Destination server    <--->      SEC-GW                                                                                       <--->Client
                                                fd5d:d50a:8c17:2010::a01<====_IPSEC channel_========>ipv6: fd5d:d50a:8c17:2010::a02
172.27.31.181 <===data===>172.27.31.175 <==_data_==>192.168.174.175<====_data_====>192.168.174.176

in this setup channel is establishing using ipv6 is working fine, but traffic via ipv4 is not working. To make it work i need to set xfrm state and traffic selection manually. but it is suppose to be set by strongswan. below are the details

SEC-GW 
```
eno1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.27.31.175  netmask 255.255.252.0  broadcast 172.27.31.255
        inet6 fdab:1234:5678:1111:681:27:31:175  prefixlen 118  scopeid 0x0<global>
        inet6 fe80::7aac:44ff:fe39:676c  prefixlen 64  scopeid 0x20<link>
        ether 78:ac:44:39:67:6c  txqueuelen 1000  (Ethernet)
        RX packets 591795  bytes 52342207 (52.3 MB)
        RX errors 0  dropped 46921  overruns 0  frame 0
        TX packets 2852  bytes 406807 (406.8 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device memory 0x9db00000-9dbfffff

eno3: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.174.175  netmask 255.255.255.0  broadcast 192.168.174.255
        inet6 fe80::7aac:44ff:fe39:676e  prefixlen 64  scopeid 0x20<link>
        inet6 fd5d:d50a:8c17:2010::a01  prefixlen 118  scopeid 0x0<global>
        ether 78:ac:44:39:67:6e  txqueuelen 1000  (Ethernet)
        RX packets 17998  bytes 2928783 (2.9 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 11595  bytes 1149347 (1.1 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device memory 0x9d900000-9d9fffff


root@ubuntu:~/# ipsec statusall
Status of IKE charon daemon (strongSwan 5.8.2, Linux 4.15.0-163-generic, x86_64):
  uptime: 114 seconds, since Jan 06 01:49:08 2022
  malloc: sbrk 2711552, mmap 0, used 638144, free 2073408
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 3
  loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp curve25519 xcbc cmac hmac drbg attr kernel-pfkey kernel-netlink resolve socket-default stroke vici updown eap-identity eap-sim eap-aka xauth-generic counters
Virtual IP pools (size/online/offline):
  192.168.174.167: 1/1/0
Listening IP addresses:
  172.27.31.175
  fdab:1234:5678:1111:681:27:31:175
  192.168.174.175
  fd5d:d50a:8c17:2010::a01
Connections:
secgw-to-client:  %any...%any6  IKEv2, dpddelay=30s
secgw-to-client:   local:  uses pre-shared key authentication
secgw-to-client:   remote: uses pre-shared key authentication
secgw-to-client:   child:  172.27.28.0/22 === dynamic TUNNEL, dpdaction=clear
Security Associations (1 up, 0 connecting):
secgw-to-client[1]: ESTABLISHED 111 seconds ago, fd5d:d50a:8c17:2010::a01[fd5d:d50a:8c17:2010::a01]...fd5d:d50a:8c17:2010::a02[fd5d:d50a:8c17:2010::a02]
secgw-to-client[1]: IKEv2 SPIs: 2c9abff597d01c96_i f9e0611s3729cb5cc_r*, rekeying in 23 hours
secgw-to-client[1]: IKE proposal: AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
secgw-to-client{1}:  INSTALLED, TUNNEL, reqid 1, ESP SPIs: c7fd1a51_i c93c40c0_o
secgw-to-client{1}:  AES_CBC_256/HMAC_SHA1_96, 0 bytes_i, 0 bytes_o, rekeying in 23 hours
secgw-to-client{1}:   172.27.28.0/22 === 192.168.174.167/32


root@ubuntu:~/# ip -s xfrm policy
src 172.27.28.0/22 dst 192.168.174.167/32 uid 0
        dir out action allow index 873 priority 372351 share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-01-06 01:49:11 use -
        tmpl src fd5d:d50a:8c17:2010::a01 dst fd5d:d50a:8c17:2010::a02
                proto esp spi 0x00000000(0) reqid 1(0x00000001) mode tunnel
                level required share any
                enc-mask 00000000 auth-mask 00000000 comp-mask 00000000
src 192.168.174.167/32 dst 172.27.28.0/22 uid 0
        dir fwd action allow index 866 priority 372351 share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-01-06 01:49:11 use -
        tmpl src fd5d:d50a:8c17:2010::a02 dst fd5d:d50a:8c17:2010::a01
                proto esp spi 0x00000000(0) reqid 1(0x00000001) mode tunnel
                level required share any
                enc-mask 00000000 auth-mask 00000000 comp-mask 00000000


ipsec.conf
conn secgw-to-client
        left=%any
        leftsubnet=172.27.31.0/22
        leftid=%any6
        rightsourceip=192.168.174.167
        right=%any6
        rightid=%any6
        authby=secret
        keyexchange=ikev2
        auto=add
        dpdaction=clear


root@ubuntu:~/# ip -s xfrm state
src fd5d:d50a:8c17:2010::a01 dst fd5d:d50a:8c17:2010::a02
        proto esp spi 0xc93c40c0(3376169152) reqid 1(0x00000001) mode tunnel
        replay-window 0 seq 0x00000006 flag  (0x00000000)
        auth-trunc hmac(sha1) 0x9dd46466a6450s352fa1659e83d1458319ee51f16 (160 bits) 96
        enc cbc(aes) 0x480bc10c3ea4d2ad1bee321bbc1a0cs091acf5950d9c31c9751353a5da76s3e6b1 (256 bits)
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        sel src ::/0 dst ::/0 uid 0
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 86063(sec), hard 86400(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-01-06 01:49:11 use -
        stats:
          replay-window 0 replay 0 failed 0
src fd5d:d50a:8c17:2010::a02 dst fd5d:d50a:8c17:2010::a01
        proto esp spi 0xc7fd1a51(3355253329) reqid 1(0x00000001) mode tunnel
        replay-window 32 seq 0x00000005 flag  (0x00000000)
        auth-trunc hmac(sha1) 0x8bd533ed9bd8sef8a6c5a5bs8b3bfee4a0cf81e22d (160 bits) 96
        enc cbc(aes) 0x6ba95c24e5519f687ad42cc5a83374sa44f33aa7bfaf6a4dd470ff6e6c721b37c (256 bits)
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        sel src ::/0 dst ::/0 uid 0
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 86098(sec), hard 86400(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-01-06 01:49:11 use -
        stats:
          replay-window 0 replay 0 failed 0
```

Client:
```
eno1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.27.31.176  netmask 255.255.252.0  broadcast 172.27.31.255
        inet6 fe80::7aac:44ff:fe39:8d54  prefixlen 64  scopeid 0x20<link>
        inet6 fdab:1234:5678:1111:681:27:31:176  prefixlen 118  scopeid 0x0<global>
        ether 78:ac:44:39:8d:54  txqueuelen 1000  (Ethernet)
        RX packets 32604  bytes 20289859 (20.2 MB)
        RX errors 0  dropped 1419  overruns 0  frame 0
        TX packets 3375  bytes 335261 (335.2 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device memory 0x9db00000-9dbfffff

eno3: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.174.176  netmask 255.255.255.0  broadcast 192.168.174.255
        inet6 fd5d:d50a:8c17:2010::a02  prefixlen 118  scopeid 0x0<global>
        inet6 fe80::7aac:44ff:fe39:8d56  prefixlen 64  scopeid 0x20<link>
        ether 78:ac:44:39:8d:56  txqueuelen 1000  (Ethernet)
        RX packets 570  bytes 118507 (118.5 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 323  bytes 42026 (42.0 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device memory 0x9d900000-9d9fffff


ipsec.conf
 conn conn-1
     leftsourceip=%config4
     left=%defaultroute
     also=profile-1
     right=fd5d:d50a:8c17:2010::a01
     leftid=%any6
     rightid=%any6
     type=tunnel
     rightsubnet=0.0.0.0/0
     auto=start
```

**Comment by tobiasbrunner** [comment]
>  ```
> src fd5d:d50a:8c17:2010::a01 dst fd5d:d50a:8c17:2010::a02
>         proto esp spi 0xc93c40c0(3376169152) reqid 1(0x00000001) mode tunnel
>         replay-window 0 seq 0x00000006 flag  (0x00000000)
>         auth-trunc hmac(sha1) 0x9dd46466a6450s352fa1659e83d1458319ee51f16 (160 bits) 96
>         enc cbc(aes) 0x480bc10c3ea4d2ad1bee321bbc1a0cs091acf5950d9c31c9751353a5da76s3e6b1 (256 bits)
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
>         sel src ::/0 dst ::/0 uid 0
> ...
>```

This looks wrong. There shouldn't be a selector for IPv6 and the `XFRM_STATE_AF_UNSPEC` flag should be set (`flag af-unspec (0x00100000)`). Have a look at this [test scenario](https://www.strongswan.org/testing/testresults/ipv6-stroke/net2net-ip4-in-ip6-ikev2/). Both should be correct automatically for tunnel mode (irrespective of the address families involved).

**Reply by SagarGowda89** [reply]
Hi Thanks for your response. i can see this is already  added in xfrm.h file.  am i missing something ? below are the file details. please check

include/linux/xfrm.h
381:#define XFRM_STATE_AF_UNSPEC        32

arm-linux-gnueabihf/include/linux/xfrm.h
381:#define XFRM_STATE_AF_UNSPEC        32

src/linux-headers-4.15.0-112/include/uapi/linux/xfrm.h
381:#define XFRM_STATE_AF_UNSPEC        32

arm-linux-gnueabi/include/linux/xfrm.h
381:#define XFRM_STATE_AF_UNSPEC        32
root@ubuntu:/usr#

**Reply by tobiasbrunner** [reply]
As I already said, this should work correctly. Why it isn't in your case I don't know.

**Reply by SagarGowda89** [reply]
yes. its not working. i have to add this manually using "ip xfrm state add". i want it to be taken automatically.

**Reply by tobiasbrunner** [reply]
You probably messed something up on your system (kernel or strongSwan installation).

**Reply by SagarGowda89** [reply]
i am following below steps for strongswan installation

    a. tar -zxvf strongswan-5.8.0.tar.gz
    b. cd strongswan-5.8.0
    c. Below configure command
        ./configure --enable-static --enable-eap-aka --enable-eap-sim --enable-eap-sim-file --enable-eap-identity --enable-kernel-pfkey --enable-openssl --enable-vici
    d. make
    e. sudo make install
    
    can i know is there anything need to be changed>

**Reply by tobiasbrunner** [reply]
No idea. Maybe you already have strongSwan binaries installed on the system (from a binary package or a previous manual installation) or there is some other issue. Also, why do you use 5.8.0 and not the latest release?

**Reply by SagarGowda89** [reply]
I have installed 5.9.4 and tried also. still issue is not resolved

**Reply by SagarGowda89** [reply]
after kernel upgrade i can new interface in clients side

eno3: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.174.176  netmask 255.255.255.0  broadcast 192.168.174.255
        inet6 fd5d:d50a:8c17:2010::a02  prefixlen 118  scopeid 0x0<global>
        inet6 fe80::7aac:44ff:fe39:8d56  prefixlen 64  scopeid 0x20<link>
        ether 78:ac:44:39:8d:56  txqueuelen 1000  (Ethernet)
        RX packets 401  bytes 83286 (83.2 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 242  bytes 30340 (30.3 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device memory 0x9d900000-9d9fffff

ipsec0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1400
        inet 192.168.174.167  netmask 255.255.255.255  destination 192.168.174.167
        inet6 fe80::b23f:1f28:70f6:35f6  prefixlen 64  scopeid 0x20<link>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 2  bytes 96 (96.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0





and i am getting below error in charan logs

unable to add SAD entry"/"Invalid argument (22)

**Reply by tobiasbrunner** [reply]
You should not have an `ipsec0` interface unless you built the [_kernel-libipsec_ plugin]([url](https://wiki.strongswan.org/projects/strongswan/wiki/kernel-libipsec)) (if this was installed by a distribution package, make sure to uninstall it). The only kernel plugin you should have on Linux is _kernel-netlink_ (_kernel-pfkey_ is useless on Linux).

**Reply by SagarGowda89** [reply]
yes i have disabled. Still issue is same . traffic selection is ipv6 only instead of ipv4


src fd5d:d50a:8c17:2010::a02 dst fd5d:d50a:8c17:2010::a01
        proto esp spi 0xc15d3f48(3244113736) reqid 1(0x00000001) mode tunnel
        replay-window 0 seq 0x00000006 flag  (0x00000000)
        auth-trunc hmac(sha1) 0x86caf08029504758a9132e6e08a968d22a74c468 (160 bits) 96
        enc cbc(aes) 0x9d6e3f4d58f9b4baba052a2dd6b31c55b531d0e9e3a2544125a40e73c82a50b9 (256 bits)
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        sel src ::/0 dst ::/0 uid 0
        lifetime config:
          limit: soft 9999999(bytes), hard 9999999(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 85435(sec), hard 86400(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-01-08 00:02:36 use -
        stats:
          replay-window 0 replay 0 failed 0
src fd5d:d50a:8c17:2010::a01 dst fd5d:d50a:8c17:2010::a02
        proto esp spi 0xc3564cad(3277212845) reqid 1(0x00000001) mode tunnel
        replay-window 32 seq 0x00000005 flag  (0x00000000)
        auth-trunc hmac(sha1) 0x1851cf56715077fc5f41a47a3047140eba39abce (160 bits) 96
        enc cbc(aes) 0x84ca9321b558b31e1c5692e87164767f0b3dd5f24d3ec4a9c3d5c91ab6a2a7f5 (256 bits)
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        sel src ::/0 dst ::/0 uid 0
        lifetime config:
          limit: soft 9999999(bytes), hard 9999999(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 85422(sec), hard 86400(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2022-01-08 00:02:36 use -
        stats:
          replay-window 0 replay 0 failed 0

**Reply by tobiasbrunner** [reply]
What plugins do you have loaded?

**Reply by SagarGowda89** [reply]
loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp curve25519 xcbc cmac hmac drbg attr kernel-pfkey kernel-netlink resolve socket-default stroke vici updown eap-identity eap-sim eap-aka xauth-generic counters

**Reply by tobiasbrunner** [reply]
Hm, what's that `--enable-` part before `kernel-pfkey`? Anyway, as already mentioned above, don't load the latter plugin, it doesn't support this kind of config (and a lot of other things either).

**Reply by SagarGowda89** [reply]
Sorry . it was by mistake while copied in to notepad and i was looking for something that time it was modified by mistake.  i have edited previous post.

**Reply by SagarGowda89** [reply]
@tobiasbrunner Issue has been solved.

Actually "kernel-pfkey" plugin was already loaded by linux, i was trying to load it in my application also. this was creating issue due to some reason instead of ignoring second loading.

**Reply by tobiasbrunner** [reply]
What do you mean? It's a plugin for strongSwan, it's never loaded by Linux. What application are you referring to? And as I said, it should never be used on Linux anyway, the plugin is for BSD-based systems.

**Reply by SagarGowda89** [reply]
when i tried to load plugin(kernel-pfkey) along with strongswan, it was throwing error in charon saying that its already loading. so loading error was there. This was causing issue. later i have removed the plugin explicit loading then issue has been resloved.

**Reply by tobiasbrunner** [reply]
OK, not sure what exactly you were trying to do (i.e. how exactly you tried to load the plugin), but great that the issue is solved.

**Reply by SagarGowda89** [reply]
@tobiasbrunner Thank you
